### ISSUE TITLE
Http messaging module

### ISSUE THREAD

**Issue by vooolll** [esl/MongooseIM#619]
Is it useful to have a module(as a pull request to current repository), which catches message text and sends it via http, on specified host and port. I have one, but I don't know may be it is not good for chat, and it's users, or insecure.

**Comment by michalwski** [comment]
Hi @vooolll,

What you are describing is used sometimes, mainly to integrate with some external chat archives. Ideally would be to come up with a solution where the HTTP endpoint is a MAM backend.

There are some security issues to consider - f.e you need to make sure the link between MongooseIM and the HTTP endpoint is secure.

**Comment by vooolll** [comment]
@michalslaski I've made it as a hook. You advice me doing it as mam backend. Like instead of storing messages in databases, it will be send over http, am I right?

**Comment by vooolll** [comment]
Actually the case here is to send via http messages one by one, but not only messages, things like "user is online"  and so on.

**Comment by michalwski** [comment]
I see, in this case it definitely doesn't sound like a MAM backend and should be done separately.
This sounds interesting and I'd really want to see the code and some use cases.

**Comment by vooolll** [comment]
The use case is again, push notifications. Some times users want to fetch the messages from chat and send(sender, receiver, text) it to their http server, so they are free to provide their own implementation of notification (email, apns, gcm etc.) It is useful because the chat is just a part of whole thing, and other is implemented via http protocol.

**Comment by michalwski** [comment]
Yes, that's true. I can see some other use cases, f.e. building a list of frequent chatters or any other stats based on the data provided to the external service.

**Comment by vooolll** [comment]
@michalwski exactly, or getting when somebody is online. My 'dirty' implementation for messaging is here https://gist.github.com/vooolll/c937ad5432f03fd8620f  it is may be configured like  ejabberd_http_auth. I will improve it.

**Comment by michalwski** [comment]
That's good starting point. Please send a PR as soon as you are ready so it's easier to comment and improve to code :)

**Comment by vooolll** [comment]
@michalwski Thanks, I will.

**Comment by vooolll** [comment]
@michalwski I am sorry for disturbing you with this kind of silly questions, have you got any idea how to test it. I actually don't understand how to specify my configuration via *.spec file(I mean I need to add something like `{{mod_http_notification, "{mod_http_notfication, [{pool_size, 5}, {host, \"localhost\"},"
      "{prefix_path, \"/test\"}]"}]}` to test.config file, while it doesn't change the ejabberd.cfg configuration. Finally I used `ejabberd.cfg` and `node1_vars.config` to create testing environment for my module, but I think it is not acceptable to send as PR. May you explain how to set up ejabberd.cfg for testing, and how to test messaging via escalus?

**Comment by michalwski** [comment]
Thanks for asking @vooolll . 

Regarding your question about config. You need to add `{{mod_http_notification}}` place holder in the template `rel/files/ejabbed.cfg` like we do for mod_private or mod_privacy f.e.

The easier way would be to start the module in the `init_per_suite` function in your new suite. There is  a helper module `dynamic_modules` which you can use to start/stop your module. With this approach you don't have to play with ejabbed.cfg nor other config files.

Regarding `escalus` - I think good example can be found in `login_SUITE` test case `messages_story`.
Basically you need to created users in `init_per_group` or `suite` and use them in `escalus:story` function.
You can also take a look at: https://github.com/esl/escalus repo for some doc.

Hope that helps.

**Comment by vooolll** [comment]
@michalwski thanks for your small guideline, the second case is what I've looked for.

**Comment by vooolll** [comment]
https://github.com/esl/MongooseIM/pull/627

**Comment by michalwski** [comment]
This feature is implemented here #684
